README.md
<img src="https://i.imgur.com/VPgx6ra.png" width="550">

# Slack Watchman for Enterprise Grid
![Python 2.7 and 3 compatible](https://img.shields.io/pypi/pyversions/slack-watchman-eg)
![PyPI version](https://img.shields.io/pypi/v/slack-watchman-eg.svg)
![License: MIT](https://img.shields.io/pypi/l/slack-watchman-eg.svg)


## About Slack Watchman for Enterprise Grid

Slack Watchman for Enterprise Grid uses the Slack Enterprise Grid DLP API to look for potentially exposed secrets and sensitive data in Slack Enterprise Grid.

**Note**: Slack Watchman for Enterprise Grid is designed for Enterprise Grid subscribers of Slack only. If you use Slack without an Enterprise subscription, you can use the standard version of [Slack Watchman](https://github.com/PaperMtn/slack-watchman)

### Features
Slack Watchman for Enterprise Grid looks for:

- API Keys, Tokens & Service Accounts
  - AWS, Azure, GCP, Google API, Slack (keys & webhooks), Twitter, Facebook, GitHub
  - Generic Private keys
  - Access Tokens, Bearer Tokens, Client Secrets, Private Tokens
- Files
  - Certificate files
  - Potentially interesting/malicious/sensitive files (.docm, .xlsm, .zip etc.)
  - Executable files
  - Keychain files
  - Config files for popular services (Terraform, Jenkins, OpenVPN and more)
- Personal Data
  - Leaked passwords
  - Passport numbers, Dates of birth, Social security numbers, National insurance numbers, Drivers licence numbers (UK), Individual Taxpayer Identification Number
  - CVs, salary information
- Financial data
  - PayPal Braintree tokens, Bank card details, IBAN numbers, CUSIP numbers
  - Budget files
  
It looks for this exposed data across all workspaces in the Enterprise, in the following locations:
- Public channels
- Private channels
- Draft messages
- Slack connect channels
- Direct messages
- Multi-person direct messages

#### Time based searching
Slack Watchman for Enterprise Grid can search through all messages sent in your Enterprise in the previous 24 hours. Limitations in the API, and data processing bottlenecks, don't allow for any further than 24 hours to be queried. 

You can provide time periods to search for using the `--hours` and `--minutes` options at runtime. This means you can schedule running regularly, and in general little and often is the best approach.

#### Multiprocessing
Multiprocessing is used to search the potentially huge amount of data retrieved when getting all messages sent in an Enterprise. You can specify how many cores to use at runtime, and the more cores you use, the faster processing is generally done. That being said, you are still constrained by the API.

I have found the most efficient approach is to use between 8-12 cores.

You can specify cores using the optional flag `--cores` at runtime. If this flag is not set, Slack Watchman will automatically use all available cores up to a maximum of 8.
### Signatures
Slack Watchman uses custom YAML signatures to detect matches in Slack. These signatures are pulled from the central [Watchman Signatures repository](https://github.com/PaperMtn/watchman-signatures). Slack Watchman for Enterprise Grid automatically updates its signature base at runtime to ensure its using the latest signatures to detect secrets. 

## Requirements
### Slack API token
To run Slack Watchman for Enterprise Grid, you will need a Slack API access token that is authorised to use the Enterprise DLP API.

To do this, you need to create a [Slack App](https://api.slack.com/apps) and install it at the organisation level.

The app needs to have the following **User Token Scopes** added:
```
discovery:read
discovery:write
team:read
users:read
```
**Note**: `discovery:read` and `discovery:write` can only be added to an app by Slack themselves, you will need to contact your Slack CSM. They will also provide you with instructions on how to install the app at organisation level and retrieve the access token.

#### Providing token
Provide the token in the environment variable `SLACK_WATCHMAN_EG_TOKEN`

## Installation
You can install the latest stable version via pip:

`python3 -m pip install slack-watchman-eg`

Or build from source yourself:

Download the release source files, then from the top level repository run:
```shell
python3 -m pip build
python3 -m pip install --force-reinstall dist/*.whl
```

## Docker Image

Slack Watchman for Enterprise Grid is also available from the Docker hub as a Docker image:

`docker pull papermountain/slack-watchman-eg:latest`

You can then run Slack Watchman for Enterprise Grid in a container, making sure you pass the required environment variables:

```
// help
docker run --rm papermountain/slack-watchman-eg -h

// scan all
docker run --rm -e SLACK_WATCHMAN_EG_TOKENN=xoxp... papermountain/slack-watchman-eg --hours 1 --cores 8
docker run --rm --env-file .env papermountain/slack-watchman-eg --hours 1 --cores 8
```

## Usage
```
usage: slack-watchman-eg [-h] [--hours HOURS] [--minutes MINUTES] [--output {json,terminal}] [--cores CORES] [--version] [--users] [--workspaces] [--debug] [--verbose]

Monitoring your Slack Enterprise Grid for sensitive information

options:
  -h, --help            show this help message and exit
  --hours HOURS, -hr HOURS
                        How far back to search in whole hours between 1-24. Defaults to 1 if no acceptable value given
  --minutes MINUTES, -m MINUTES
                        How far back to search in whole minutes between 1-60
  --output {json,terminal}, -o {json,terminal}
                        What logging output to use - JSON formatted output, or textual outputfor reading via terminal. Default is terminal
  --cores CORES, -c CORES
                        Number of cores to use between 1-12
  --version, -v         show program's version number and exit
  --users, -u           Return all users
  --workspaces, -w      Return all workspaces
  --debug, -d           Turn on debug level logging
  --verbose, -V         Turn on more verbose output for JSON logging. This includes more fields, but is larger
```

## Other Watchman apps
You may be interested in the other apps in the Watchman family:
- [Slack Watchman](https://github.com/PaperMtn/slack-watchman)
- [GitLab Watchman](https://github.com/PaperMtn/gitlab-watchman)
- [GitHub Watchman](https://github.com/PaperMtn/github-watchman)

## License
The source code for this project is released under the [GNU General Public Licence](https://www.gnu.org/licenses/licenses.html#GPL). This project is not associated with Slack Technologies or Salesforce.

CHANGELOG.md
## [2.0.0] - 2023-04-14
This major version release brings multiple updates to Slack Watchman for Enterprise Grid, both in usability, functionality and behind the scenes improvements.
### Added
- Support for centralised signatures from the Watchman Signatures repository. This makes it much easier to keep the signature base for all Watchman applications up to date, and to add functionality to Slack Watchman with new signatures. New signatures are downloaded, and updates to existing signatures are applied, at runtime, meaning Slack Watchman for Enterprise Grid will always be using the most up to date signatures. 
- Option for terminal optimised logging instead of JSON formatting. This is now the default when running with no output option selected, and is a lot easier for humans to read. Also, colours! 
- Option choose between verbose or succinct logging when using JSON output. Default is succinct.
- Debug logging option
### Removed
- Support for tombstoning posts that match signatures removed
- Local signatures - Centralised signatures mean that user-created custom signatures can't be used with Slack Watchman for Enterprise Grid anymore. If you have made a signature you think would be good for sharing with the community, feel free to add it to the Watchman Signatures repository, so it can be used in all Watchman applications 
- For the reason above, the functionality to have sandbox signatures has been removed as well
### Fixed
- Draft searches were giving an error due to not being able to populate some workspace information. This has now been fixed

## [1.1.1] - 2022-05-16
### Added
- Signature to find Atlassian tokens

## [1.1.0] - 2022-04-02
### Added
- Docker image now available from the Docker hub, or by building from source.
- Support for Python 3.7
- New logo to play nicely with dark mode
### Fixed
- More errors when importing packages

## [1.0.2] - 2021-12-30
### Fixed
- Error when importing packages
- Signatures not being included in the distribution package

## [1.0.1] - 2021-12-30
### Added
- Refactor and update distribution files

## [1.0.0] - 2021-12-30
- Initial release



src/slack_watchman_eg/__init__.py
import argparse
import multiprocessing
import os
import time
import calendar
from pathlib import Path
from typing import List

from . import __version__
from . import sw_logger
from . import slack_wrapper
from . import signature_updater
from .models import (
    signature,
    user,
    workspace,
    post,
    conversation
)

SIGNATURES_PATH = (Path(__file__).parents[2] / 'watchman-signatures').resolve()
OUTPUT_LOGGER: sw_logger.JSONLogger


def load_signatures() -> List[signature.Signature]:
    """ Load signatures from YAML files
    Returns:
        List containing loaded definitions as Signatures objects
    """

    loaded_signatures = []
    try:
        for root, dirs, files in os.walk(SIGNATURES_PATH):
            for sig_file in files:
                sig_path = (Path(root) / sig_file).resolve()
                if sig_path.name.endswith('.yaml'):
                    loaded_def = signature.load_from_yaml(sig_path)
                    for sig in loaded_def:
                        if sig.status == 'enabled' and 'slack_eg' in sig.watchman_apps:
                            loaded_signatures.append(sig)
        return loaded_signatures
    except Exception as e:
        raise e


def search(loaded_signature: signature,
           slack_connection: slack_wrapper.SlackAPI,
           user_list: List[user.User],
           message_list: List[post.Message],
           workspace_list: List[workspace.Workspace],
           files_list: List[post.File],
           scope: str,
           cores: int,
           verbose: bool) -> None:
    """ Uses the signature to call the relevant search functions to find data in messages
    files and drafts. Results are output to stdout logging.
    Args:
        slack_connection: Slack API object
        loaded_signature: Signature object
        user_list: List of User objects from the Enterprise
        message_list: List of Message objects to search through
        workspace_list: List of Workspace objects from the Enterprise
        files_list: List of File objects to search through
        scope: Scope of any results found for logging: e.g. Draft
        cores: Number of CPU cores to use
        verbose: Whether to use verbose logging or not
    """

    if scope == 'messages':
        OUTPUT_LOGGER.log('INFO', f'Searching for posts containing {loaded_signature.name}')
        messages = slack_wrapper.search_message_matches(
            loaded_signature,
            slack_connection,
            user_list,
            message_list,
            workspace_list,
            cores,
            OUTPUT_LOGGER,
            verbose
        )
        if messages:
            for message in messages:
                OUTPUT_LOGGER.log(
                    'NOTIFY',
                    message,
                    scope=scope,
                    severity=loaded_signature.severity,
                    detect_type=loaded_signature.name,
                    notify_type='result'
                )

    if scope == 'files':
        OUTPUT_LOGGER.log('INFO', f'Searching for {loaded_signature.name}')
        files = slack_wrapper.search_file_matches(
            loaded_signature,
            user_list,
            files_list,
            cores,
            OUTPUT_LOGGER
        )
        if files:
            for file in files:
                OUTPUT_LOGGER.log(
                    'NOTIFY',
                    file,
                    scope=scope,
                    severity=loaded_signature.severity,
                    detect_type=loaded_signature.name,
                    notify_type='result'
                )


def core_validation(cores: int) -> int:
    """ Validate the number of cores entered
    Args:
        cores: Number of cores passed to CLI
    Returns:
        Number of cores to use
    """

    if not cores or cores > multiprocessing.cpu_count() or cores > 12:
        if multiprocessing.cpu_count() >= 8:
            return 8
        else:
            return multiprocessing.cpu_count()
    else:
        return cores


def init_logger(logging_type: str, debug: bool) -> sw_logger.JSONLogger or sw_logger.StdoutLogger:
    """ Create a logger object
    Returns:
        Logging object for outputting results
    """

    if not logging_type or logging_type == 'terminal':
        return sw_logger.StdoutLogger(debug=debug)
    else:
        return sw_logger.JSONLogger(debug=debug)


def main():
    global OUTPUT_LOGGER
    try:
        OUTPUT_LOGGER = ''
        parser = argparse.ArgumentParser(description=__version__.__summary__)

        parser.add_argument('--hours', '-hr', dest='hours', type=int,
                            help='How far back to search in whole hours between 1-24. Defaults to 1 if no acceptable '
                                 'value given', required=False)
        parser.add_argument('--minutes', '-m', dest='minutes', type=int,
                            help='How far back to search in whole minutes between 1-60', required=False)
        parser.add_argument('--output', '-o', choices=['json', 'terminal'], dest='logging_type',
                            help='What logging output to use - JSON formatted output, or textual output'
                                 'for reading via terminal. Default is terminal')
        parser.add_argument('--cores', '-c', dest='cores', type=int,
                            help='Number of cores to use between 1-12', required=False)
        parser.add_argument('--version', '-v', action='version',
                            version=f'Slack Watchman for Enterprise Grid: {__version__.__version__}')
        parser.add_argument('--users', '-u', dest='users', action='store_true', help='Return all users')
        parser.add_argument('--workspaces', '-w', dest='workspaces', action='store_true', help='Return all workspaces')
        parser.add_argument('--debug', '-d', dest='debug', action='store_true', help='Turn on debug level logging')
        parser.add_argument('--verbose', '-V', dest='verbose', action='store_true',
                            help='Turn on more verbose output for JSON logging. '
                                 'This includes more fields, but is larger')

        args = parser.parse_args()
        hours = args.hours
        minutes = args.minutes
        cores = args.cores
        users = args.users
        workspaces = args.workspaces
        logging_type = args.logging_type
        debug = args.debug
        verbose = args.verbose

        span = 0
        OUTPUT_LOGGER = init_logger(logging_type, debug)
        if minutes:
            if isinstance(minutes, int) and 1 <= int(minutes) <= 60:
                span = (int(minutes) * 60)
            else:
                minutes = None

        if hours:
            if isinstance(hours, int) and 1 <= int(hours) <= 24:
                span = (span + (hours * 3600))
            else:
                hours = None

        if not minutes and not hours:
            tf = calendar.timegm(time.gmtime()) - 3600
            hours = 1
            minutes = 0
        else:
            tf = calendar.timegm(time.gmtime()) - span

        if not minutes:
            minutes = 0
        if not hours:
            hours = 0

        cores = core_validation(cores)

        slack_con = slack_wrapper.initiate_slack_connection(os.environ.get('SLACK_WATCHMAN_EG_TOKEN'))
        OUTPUT_LOGGER.log('SUCCESS', 'Slack Watchman Enterprise Grid started execution')
        OUTPUT_LOGGER.log('INFO', f'Version: {__version__.__version__}')
        OUTPUT_LOGGER.log('INFO', f'Created by: {__version__.__author__} - {__version__.__email__}')
        OUTPUT_LOGGER.log('INFO', f'{cores} cores in use')
        OUTPUT_LOGGER.log('INFO', 'Downloading signature file updates')
        signature_updater.SignatureUpdater(OUTPUT_LOGGER).update_signatures()
        OUTPUT_LOGGER.log('INFO', 'Importing signatures...')
        signature_list = load_signatures()
        OUTPUT_LOGGER.log('SUCCESS', f'{len(signature_list)} signatures loaded')
        OUTPUT_LOGGER.log('INFO', f'Searching previous {hours} hour(s), {minutes} minutes')
        OUTPUT_LOGGER.log('INFO', 'Enumerating Enterprise information')
        OUTPUT_LOGGER.log('NOTIFY', slack_wrapper.get_enterprise(slack_con), scope='Enterprise',
                          notify_type='enterprise')
        OUTPUT_LOGGER.log('INFO', 'Enumerating Enterprise workspaces')
        workspace_list = slack_wrapper.get_workspaces(slack_con, verbose)
        OUTPUT_LOGGER.log('INFO', f'{len(workspace_list)} workspaces discovered')
        OUTPUT_LOGGER.log('INFO', 'Enumerating Enterprise users')
        user_list = slack_wrapper.get_users(slack_con, workspace_list, verbose)
        OUTPUT_LOGGER.log('INFO', f'{len(user_list)} users discovered')

        if users:
            OUTPUT_LOGGER.log('INFO', 'Outputting Enterprise users')
            for user in user_list:
                OUTPUT_LOGGER.log('NOTIFY', user, detect_type='User', notify_type='user')

        if workspaces:
            OUTPUT_LOGGER.log('INFO', 'Outputting Enterprise workspaces')
            for workspace in workspace_list:
                OUTPUT_LOGGER.log('NOTIFY', workspace, detect_type='Workspace', notify_type='workspace')

        OUTPUT_LOGGER.log('INFO', 'Enumerating files')
        file_list = slack_wrapper.get_all_files(slack_con, cores=cores, verbose=verbose, timeframe=tf)
        OUTPUT_LOGGER.log('INFO', f'{len(file_list)} files discovered')

        OUTPUT_LOGGER.log('INFO', 'Enumerating messages')
        message_list = slack_wrapper.get_all_messages(slack_con, cores=cores, timeframe=tf)
        OUTPUT_LOGGER.log('INFO', f'{len(message_list)} messages discovered')

        for sig in signature_list:
            for scope in sig.scope:
                search(
                    sig,
                    slack_con,
                    user_list,
                    message_list,
                    workspace_list,
                    file_list,
                    scope,
                    cores,
                    verbose
                )

        OUTPUT_LOGGER.log('INFO', 'Enumerating draft messages')
        draft_list = slack_wrapper.get_all_drafts(
            slack_con,
            workspace_list,
            cores=cores,
            verbose=verbose,
            timeframe=tf
        )
        OUTPUT_LOGGER.log('INFO', f'{len(draft_list)} drafts discovered')
        for sig in signature_list:
            if 'drafts' in sig.scope:
                OUTPUT_LOGGER.log('INFO', f'Searching for drafts containing {sig.name}')
                drafts = slack_wrapper.search_draft_matches(
                    slack_con,
                    sig,
                    draft_list,
                    user_list,
                    OUTPUT_LOGGER,
                    verbose,
                    tf
                )
                if drafts:
                    for draft in drafts:
                        OUTPUT_LOGGER.log(
                            'NOTIFY',
                            draft,
                            scope='Draft',
                            severity=sig.severity,
                            detect_type=sig.name,
                            notify_type='result'
                        )

        OUTPUT_LOGGER.log('SUCCESS', 'Slack Watchman Enterprise Grid finished execution')

    except Exception as e:
        OUTPUT_LOGGER.log('CRITICAL', e)


if __name__ == '__main__':
    main()
